codelabs/custom_generator/custom_generator.md

```diff
@@ -0,0 +1,379 @@
+
+### What you'll learn
+- How to create a custom language generator.
+- How to create block generator definitions for existing blocks.
+- How to create block generator definitions for new block.
+- How to use a custom generator in the playground.
+
+### What you'll build
+
+You will build a JSON generator that implements the [JSON language spec](https://www.json.org/json-en.html).
+
+![](./json_workspace.png)
+
+### What you'll need
+
+- Familiarity with JSON and the JSON specification.
+- Comfort with defining blocks and toolboxes.
+
+This codelab assumes that you are already comfortable with using the Blockly playground locally.  You can find it in `tests/playground.html`.
+
+## Setup
+
+In this codelab you will add code to the Blockly playground to create and use a new generator. The playground contains all of Blockly's base blocks. You can find the playground at `tests/playground.html`.
+
+To start, create a file named `custom_generator.js` in the same folder as the playground.  Include it with a script tag.
+
+```
+<script src="custom_generator.js"></script>
+```
+
+Note: you must include your custom code *after* including the Blockly library.
+
+### Blocks
+
+For this codelab you will use two custom blocks, as well as five blocks from Blockly's standard set.
+
+The custom blocks represent the *Object* and *Member* sections of the JSON specification.
+
+The blocks are:
+- object
+- member
+- math_number
+- text
+- logic_boolean
+- logic_null
+- lists_create_with
+
+### Custom block definitions
+
+Copy this code into `custom_generator.js` to define the two custom blocks.
+
+```js
+Blockly.defineBlocksWithJsonArray([{
+  "type": "object",
+  "message0": "{ %1 %2 }",
+  "args0": [
+    {
+      "type": "input_dummy"
+    },
+    {
+      "type": "input_statement",
+      "name": "MEMBERS"
+    }
+  ],
+  "output": null,
+  "colour": 230,
+},
+{
+  "type": "member",
+  "message0": "%1 %2 %3",
+  "args0": [
+    {
+      "type": "field_input",
+      "name": "MEMBER_NAME",
+      "text": ""
+    },
+    {
+      "type": "field_label",
+      "name": "COLON",
+      "text": ":"
+    },
+    {
+      "type": "input_value",
+      "name": "MEMBER_VALUE"
+    }
+  ],
+  "previousStatement": null,
+  "nextStatement": null,
+  "colour": 230,
+}]);
+```
+
+### Toolbox definition
+
+Next, define your toolbox in XML. For this example we have a flyout-only toolbox with seven blocks in it.
+
+Copy this code into `custom_generator.js`:
+
+```js
+var codelabToolbox = `
+<xml id="toolbox" style="display: none">
+<block type="object"/>
+<block type="member"></block>
+<block type="math_number"><field name="NUM">0</field></block>
+<block type="text"><field name="TEXT"/></block>
+<block type="logic_boolean"><field name="BOOL">TRUE</field></block>
+<block type="logic_null"/>
+<block type="lists_create_with"><mutation items="3"/></block></xml>
+`
+```
+
+Then update the options struct in the playgound to use your new toolbox:
+
+```js
+var defaultOptions = {
```

Comment:
Given you're using the playground.html for this codelab, I wonder if: 
- we update our playground.html to have a defaultOptions struct rather than as it is right now (inline)
- update the codelab to use the advanced playground
- (just a thought) maybe we build / use an interactive environment similar to https://observablehq.com/@samelhusseini/blockly-notebook

Reply:
That's one of my big questions: where do I tell them to set this up?
I'm leaning toward using the advanced playground so I can get the benefits of our dev tools. The only downside I see is that I might have to update this codelab shortly after publishing it, especially if your playground work gets to the point where it basically replaces the advanced playground.

This same issue shows up where I say to add a button to generate code: if your playground includes setups for adding new generators, do I need to have code for them to add a button? So far I've just been logging to the console, but really I should use the text area instead.

re: observablehq: ooh, that's nifty

Reply:
Agreed that we need a consistent strategy for where one follows along. The advanced playground, aside from its name (might be unfriendly to beginners), makes sense to me as we can setup things like mechanisms for registering a new generator, adding a button, etc.